tls/ttls.c

```diff
@@ -2226,7 +2228,7 @@ ttls_recv(void *tls_data, unsigned char *buf, size_t len, unsigned int *read)
 	}
 	WARN_ON_ONCE(!io->msglen);
 	delta = *read - parsed;
-	if (delta == len)
+	if (delta == len && *read < io->msglen)
```

Comment:
`msglen` doesn't include handshake header size, so don't we need to account for the header size in the condition? Also this is the second fix in the piece of code for the last weeks - the piece is tricky, so please leave a comment in which cases, after header parsing a new record rf reentering the function for the next chunk of a record, we return `TFW_POSTPONE`.

Reply:
You are right, we need to add TLS_HEADER_SIZE to the `io->msglen` here.
Moreover the way I tried to use `*read` is not right at all. Its value has no sense, only difference matter. That's why `delta` is computed as a delta. But we don't need a delta, we need an absolute value. At first I was going to change condition to `io->rlen + len < io->msglen + TLS_HEADER_SIZE` and that looks kind of OK. There is even similar condition `io->msglen > io->rlen + len` later in the code.

But at the same time, `ttls_parse_record_hdr()` which is called a bit earlier, is changing `io->rlen`, for application data records and for alerts, but latter only if `ttls_xfrm_ready(tls)` returns truth. So `io->rlen + len` cannot be used for the purpose of measuring amount of data read.

Now I doubt the `io->msglen > io->rlen + len` condition is the right one. Either there is a trick of some kind which I don't get yet, or the condition is actually wrong.